src/serac/physics/integrators/traction_integrator.cpp

```diff
@@ -0,0 +1,226 @@
+
+    // Compute the traction at the integration point
+    traction_.Eval(traction_vector_, *parent_to_reference_face_transformation.Face, ip);
+
+    // If computing on the deformed configuration, transform the normal vector appropriately
+    if (!compute_on_reference_) {
+      // Calculate the deformation gradient and its inverse
+      parent_to_reference_face_transformation.Elem1->SetIntPoint(&eip);
+      CalcInverse(parent_to_reference_face_transformation.Elem1->Jacobian(), dxi_dX_);
+      element_1.CalcDShape(eip, dN_dxi_);
+      Mult(dN_dxi_, dxi_dX_, dN_dX_);
+      MultAtB(input_state_matrix_, dN_dX_, F_);
+
+      for (int d = 0; d < dim; d++) {
+        F_(d, d) += 1.0;
+      }
+
+      CalcInverse(F_, Finv_);
+
+      // Calculate the normal vector in the reference configuration
+      CalcOrtho(parent_to_reference_face_transformation.Face->Jacobian(), reference_normal_);
+
+      // Normalize vector
+      double norm = reference_normal_.Norml2();
+      reference_normal_ /= norm;
+
+      // Calculate F inverse time the normal vector
+      Finv_.MultTranspose(reference_normal_, current_normal_);
+    }
+
+    // Calculate the shape functions
+    element_1.CalcShape(eip, shape_);
+    for (int j = 0; j < dof; j++) {
+      for (int k = 0; k < dim; k++) {
+        double residual_contribution =
+            -1.0 * traction_vector_(k) * shape_(j) * ip.weight * parent_to_reference_face_transformation.Face->Weight();
+        if (!compute_on_reference_) {
+          residual_contribution *= F_.Det() * current_normal_.Norml2();
+        }
+        output_residual_vector(dof * k + j) += residual_contribution;
+      }
+    }
+  }
+}
+
+void TractionIntegrator::AssembleFaceGrad(const mfem::FiniteElement& element_1, const mfem::FiniteElement& element_2,
+                                          mfem::FaceElementTransformations& parent_to_reference_face_transformation,
+                                          const mfem::Vector& input_state_vector, mfem::DenseMatrix& stiffness_matrix)
+{
+  double       diff_step = 1.0e-8;
+  mfem::Vector temp_out_1;
+  mfem::Vector temp_out_2;
+  mfem::Vector temp(input_state_vector.GetData(), input_state_vector.Size());
+
+  stiffness_matrix.SetSize(input_state_vector.Size(), input_state_vector.Size());
+  stiffness_matrix = 0.0;
+
+  // If computing on the deformed configuration calculate the stiffness contributions via finite difference
+  // Otherwise, the contributuion is zero
+  if (!compute_on_reference_) {
+    for (int j = 0; j < temp.Size(); j++) {
+      temp[j] += diff_step;
+      AssembleFaceVector(element_1, element_2, parent_to_reference_face_transformation, temp, temp_out_1);
+      temp[j] -= 2.0 * diff_step;
+      AssembleFaceVector(element_1, element_2, parent_to_reference_face_transformation, temp, temp_out_2);
+
+      for (int k = 0; k < temp.Size(); k++) {
+        stiffness_matrix(k, j) = (temp_out_1[k] - temp_out_2[k]) / (2.0 * diff_step);
```

Comment:
The routine is called "facegrad" but here is the stiffness matrix.  But maybe better call it the tangent matrix

Reply:
MFEM-ism.